File: src/sc_neurocore/neurons/models/sc_normalized_energy_lif.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Retained normalized energy-gated exact-flow LIF

"""SC compatibility model for the former EnergyLIF recurrence."""

from __future__ import annotations

import math
from dataclasses import dataclass

_VOLTAGE_MIN = -200.0
_VOLTAGE_MAX = 100.0
_ENERGY_GATE = 0.1


@dataclass
class SCNormalizedEnergyLIFNeuron:
    """Retain the project's normalized two-state energy-gated LIF exactly."""

    v: float = -70.0
    epsilon: float = 1.0
    v_rest: float = -70.0
    v_reset: float = -70.0
    v_threshold: float = -50.0
    tau_m: float = 10.0
    tau_e: float = 500.0
    alpha: float = 0.1
    epsilon_0: float = 1.0
    resistance: float = 1.0
    dt: float = 1.0

    def __post_init__(self) -> None:
        """Validate the retained SC state before first use."""
        self._validate_state()

    def _validate_state(self) -> None:
        """Reject invalid retained SC state before mutation."""
        for field in ("v", "v_rest", "v_reset", "v_threshold"):
            if not math.isfinite(getattr(self, field)):
                raise ValueError(f"{field} must be finite")
        for field in ("epsilon", "epsilon_0"):
            value = getattr(self, field)
            if not math.isfinite(value) or value < 0.0:
                raise ValueError(f"{field} must be finite and non-negative")
        for field in ("tau_m", "tau_e", "resistance", "dt"):
            value = getattr(self, field)
            if not math.isfinite(value) or value <= 0.0:
                raise ValueError(f"{field} must be finite and positive")
        if not math.isfinite(self.alpha) or self.alpha < 0.0:
            raise ValueError("alpha must be finite and non-negative")
        if self.epsilon > self.epsilon_0:
            raise ValueError("epsilon must not exceed epsilon_0")
        if not _VOLTAGE_MIN <= self.v <= _VOLTAGE_MAX:
            raise ValueError("v must be inside the voltage safety envelope")
        if not _VOLTAGE_MIN <= self.v_reset <= _VOLTAGE_MAX:
            raise ValueError("v_reset must be inside the voltage safety envelope")
        if self.dt > self.tau_m or self.dt > self.tau_e:
            raise ValueError("dt must not exceed tau_m or tau_e")
        if self.v_threshold <= self.v_rest or self.v_threshold <= self.v_reset:
            raise ValueError("v_threshold must exceed resting and reset voltage")

    def _exact_candidate(self, current: float) -> tuple[float, float]:
        """Return the retained exact constant-current candidate."""
        membrane_decay = math.exp(-self.dt / self.tau_m)
        energy_decay = math.exp(-self.dt / self.tau_e)
        energy_delta = self.epsilon - self.epsilon_0
        epsilon_candidate = self.epsilon_0 + energy_delta * energy_decay
        steady = self.epsilon_0 * self.tau_m * (1.0 - membrane_decay)
        coupled_rate = (1.0 / self.tau_m) - (1.0 / self.tau_e)
        transient = (
            energy_delta * membrane_decay * self.dt
            if abs(coupled_rate) < 1.0e-12
            else energy_delta * membrane_decay * math.expm1(coupled_rate * self.dt) / coupled_rate
        )
        v_candidate = (
            self.v_rest
            + (self.v - self.v_rest) * membrane_decay
            + (self.resistance * current / self.tau_m) * (steady + transient)
        )
        return v_candidate, epsilon_candidate

    def step(self, current: float) -> int:
        """Advance one retained exact-flow sample and return its event."""
        if not math.isfinite(current):
            raise ValueError("current must be finite")
        self._validate_state()
        v_candidate, epsilon_candidate = self._exact_candidate(current)
        if not (
            math.isfinite(v_candidate)
            and _VOLTAGE_MIN <= v_candidate <= _VOLTAGE_MAX
            and math.isfinite(epsilon_candidate)
            and 0.0 <= epsilon_candidate <= self.epsilon_0
        ):
            raise ValueError("SC energy-LIF candidate left the safety envelope")
        if v_candidate >= self.v_threshold and epsilon_candidate > _ENERGY_GATE:
            self.v = self.v_reset
            self.epsilon = max(0.0, epsilon_candidate - self.alpha)
            return 1
        self.v = v_candidate
        self.epsilon = epsilon_candidate
        return 0

    def reset(self) -> None:
        """Restore the retained normalized resting state."""
        self.v = self.v_rest
        self.epsilon = self.epsilon_0

File: engine/src/neurons/trivial/sc_normalized_energy_lif.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li

//! Retained normalized energy-gated exact-flow SC model.

const V_MIN: f64 = -200.0;
const V_MAX: f64 = 100.0;

#[derive(Clone, Debug)]
pub struct SCNormalizedEnergyLIFNeuron {
    pub v: f64,
    pub epsilon: f64,
    pub v_rest: f64,
    pub v_reset: f64,
    pub v_threshold: f64,
    pub tau_m: f64,
    pub tau_e: f64,
    pub alpha: f64,
    pub epsilon_0: f64,
    pub resistance: f64,
    pub dt: f64,
}

impl SCNormalizedEnergyLIFNeuron {
    pub fn new() -> Self {
        Self {
            v: -70.0,
            epsilon: 1.0,
            v_rest: -70.0,
            v_reset: -70.0,
            v_threshold: -50.0,
            tau_m: 10.0,
            tau_e: 500.0,
            alpha: 0.1,
            epsilon_0: 1.0,
            resistance: 1.0,
            dt: 1.0,
        }
    }
    pub fn valid(&self) -> bool {
        [
            self.v,
            self.epsilon,
            self.v_rest,
            self.v_reset,
            self.v_threshold,
            self.tau_m,
            self.tau_e,
            self.alpha,
            self.epsilon_0,
            self.resistance,
            self.dt,
        ]
        .into_iter()
        .all(f64::is_finite)
            && (V_MIN..=V_MAX).contains(&self.v)
            && (V_MIN..=V_MAX).contains(&self.v_reset)
            && (0.0..=self.epsilon_0).contains(&self.epsilon)
            && self.epsilon_0 >= 0.0
            && self.alpha >= 0.0
            && self.tau_m > 0.0
            && self.tau_e > 0.0
            && self.resistance > 0.0
            && self.dt > 0.0
            && self.dt <= self.tau_m
            && self.dt <= self.tau_e
            && self.v_threshold > self.v_rest
            && self.v_threshold > self.v_reset
    }
    pub fn try_step(&mut self, current: f64) -> Result<i32, &'static str> {
        if !self.valid() || !current.is_finite() {
            return Err("invalid SC normalized EnergyLIF state, configuration, or current");
        }
        let md = (-self.dt / self.tau_m).exp();
        let ed = (-self.dt / self.tau_e).exp();
        let de = self.epsilon - self.epsilon_0;
        let epsilon = self.epsilon_0 + de * ed;
        let steady = self.epsilon_0 * self.tau_m * (1.0 - md);
        let rate = 1.0 / self.tau_m - 1.0 / self.tau_e;
        let transient = if rate.abs() < 1.0e-12 {
            de * md * self.dt
        } else {
            de * md * (rate * self.dt).exp_m1() / rate
        };
        let v = self.v_rest
            + (self.v - self.v_rest) * md
            + (self.resistance * current / self.tau_m) * (steady + transient);
        if !(v.is_finite()
            && (V_MIN..=V_MAX).contains(&v)
            && epsilon.is_finite()
            && (0.0..=self.epsilon_0).contains(&epsilon))
        {
            return Err("SC normalized EnergyLIF candidate outside safety envelope");
        }
        if v >= self.v_threshold && epsilon > 0.1 {
            self.v = self.v_reset;
            self.epsilon = (epsilon - self.alpha).max(0.0);
            return Ok(1);
        }
        self.v = v;
        self.epsilon = epsilon;
        Ok(0)
    }
    pub fn step(&mut self, current: f64) -> i32 {
        self.try_step(current).unwrap_or(-1)
    }
    pub fn reset(&mut self) {
        self.v = self.v_rest;
        self.epsilon = self.epsilon_0;
    }
}
impl Default for SCNormalizedEnergyLIFNeuron {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn retained_trace_is_frozen() {
        let mut n = SCNormalizedEnergyLIFNeuron::new();
        let drive = [30.0, 0.0, 50.0, 10.0];
        let events: i32 = (0..256).map(|i| n.step(drive[i % drive.len()])).sum();
        assert_eq!(events, 3);
        assert!((n.v - (-52.508269792668216)).abs() < 1e-12);
    }
}
